atlasdb-config/src/main/java/com/palantir/atlasdb/factory/timelock/TimestampCorroboratingTimelockService.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.factory.timelock;
+
+import java.util.function.Supplier;
+import java.util.function.ToLongFunction;
+
+import com.palantir.lock.v2.AutoDelegate_TimelockService;
+import com.palantir.lock.v2.IdentifiedTimeLockRequest;
+import com.palantir.lock.v2.StartAtlasDbTransactionResponse;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.timestamp.TimestampRange;
+
+/**
+ * A timelock service decorator for introducing runtime validity checks on received timestamps.
+ */
+public class TimestampCorroboratingTimelockService implements AutoDelegate_TimelockService {
+    static final String CLOCKS_WENT_BACKWARDS_MESSAGE =
+            "Expected timestamp to be greater than %s, but a fresh timestamp was %s!";
+
+    private final TimelockService delegate;
+    private volatile long lowerBound = Long.MIN_VALUE;
```

Comment:
please use a final LongAccumulator for this (new LongAccumulator(Math::max, Long.MIN_VALUE)).

Reply:
basically you should never need 'synchronized' or 'volatile' in this class